homeassistant/components/axis/switch.py

```diff
@@ -1,46 +1,88 @@
 from homeassistant.core import HomeAssistant, callback
 from homeassistant.helpers.entity_platform import AddEntitiesCallback
 
 from .entity import AxisEventEntity
 from .hub import AxisHub
 
 
+@dataclass(frozen=True, kw_only=True)
+class AxisSwitchDescription(SwitchEntityDescription):
+    """Axis switch entity description."""
+
+    event_topic: EventTopic
+    """Event topic that provides state updates."""
+    name_fn: Callable[[AxisHub, Event], str]
+    """Function providing the corresponding name to the event ID."""
+    supported_fn: Callable[[AxisHub, Event], bool]
+    """Function validating if event is supported."""
+
+
+ENTITY_DESCRIPTIONS = (
+    AxisSwitchDescription(
+        key="Relay state control",
+        device_class=SwitchDeviceClass.OUTLET,
+        entity_category=EntityCategory.CONFIG,
+        event_topic=EventTopic.RELAY,
+        supported_fn=lambda hub, event: isinstance(int(event.id), int),
+        name_fn=lambda hub, event: hub.api.vapix.ports[event.id].name,
+    ),
+)
+
+
 async def async_setup_entry(
     hass: HomeAssistant,
     config_entry: ConfigEntry,
     async_add_entities: AddEntitiesCallback,
 ) -> None:
-    """Set up a Axis switch."""
+    """Set up the Axis switch platform."""
     hub = AxisHub.get_hub(hass, config_entry)
 
     @callback
-    def async_create_entity(event: Event) -> None:
-        """Create Axis switch entity."""
-        async_add_entities([AxisSwitch(event, hub)])
+    def register_platform(descriptions: Iterable[AxisSwitchDescription]) -> None:
+        """Register entity platform to create entities on event initialized signal."""
+
+        @callback
+        def create_entity(description: AxisSwitchDescription, event: Event) -> None:
+            """Create Axis entity."""
+            if description.supported_fn(hub, event):
+                async_add_entities([AxisSwitch(hub, description, event)])
+
+        for description in descriptions:
+            hub.api.event.subscribe(
+                partial(create_entity, description),
+                topic_filter=description.event_topic,
+                operation_filter=EventOperation.INITIALIZED,
+            )
 
-    hub.api.event.subscribe(
-        async_create_entity,
-        topic_filter=EventTopic.RELAY,
-        operation_filter=EventOperation.INITIALIZED,
-    )
+    register_platform(ENTITY_DESCRIPTIONS)
 
 
 class AxisSwitch(AxisEventEntity, SwitchEntity):
     """Representation of a Axis switch."""
 
-    def __init__(self, event: Event, hub: AxisHub) -> None:
+    def __init__(
+        self, hub: AxisHub, description: AxisSwitchDescription, event: Event
+    ) -> None:
         """Initialize the Axis switch."""
         super().__init__(event, hub)
-        if event.id and hub.api.vapix.ports[event.id].name:
-            self._attr_name = hub.api.vapix.ports[event.id].name
+        self.entity_description = description
+        self._attr_name = description.name_fn(hub, event) or self._attr_name
```

Comment:
name_fn always returns a str, why do we "or self.atrrname`?

Reply:
So the name can be "" then we fallback to the name based on event data provided by base class

Reply:
And what can that name be? (And do we have tests for that?)

Reply:
Test yes

Mock data https://github.com/home-assistant/core/blob/2a5c85a020e22022a82ac9a8508be5e0eaa03b12/tests/components/axis/test_switch.py#L43

Validation https://github.com/home-assistant/core/blob/2a5c85a020e22022a82ac9a8508be5e0eaa03b12/tests/components/axis/test_switch.py#L73

Reply:
Base class
https://github.com/home-assistant/core/blob/2a5c85a020e22022a82ac9a8508be5e0eaa03b12/homeassistant/components/axis/entity.py#L79